### PR TITLE
feat: support maxmemory-clients config

### PR DESCRIPTION
**Pull request by AntiTopQuark**
relate to https://github.com/apache/kvrocks/issues/2284
Implemented a new maxmemory-clients configuration to enable client eviction based on overall memory usage, preventing potential OOM issues due to excessive client connections.

### CODE REVIEW

File: src/server/server.cc [hunk 1]
```diff
@@ -1136,12 +1141,13 @@ void Server::GetStatsInfo(std::string *info) {
...
   string_stream << "sync_partial_ok:" << stats.psync_ok_count << "\r\n";
   string_stream << "sync_partial_err:" << stats.psync_err_count << "\r\n";
+  string_stream << "evicted_clients:" << stats.stat_evictedclients << "\r\n";
 
   auto db_stats = storage->GetDBStats();
   string_stream << "keyspace_hits:" << db_stats->keyspace_hits << "\r\n";
   string_stream << "keyspace_misses:" << db_stats->keyspace_misses << "\r\n";
 
-  {
+  {https://www.anyknew.com/go/10112884
```
Comment (PragmaTwice) [inline src/server/server.cc]: hmm

Comment (AntiTopQuark) [reply src/server/server.cc]: I'm sorry for the mistake

Comment (AntiTopQuark) [reply src/server/server.cc]: I'll review the code again shortly.

**Comment by caipengbo** [comment]
I want to discuss how we can implement this feature. My idea is to count client memory within each worker (using `libevent`'s callback to count `evbuffer`) to avoid locking across workers.  @AntiTopQuark @PragmaTwice @git-hulk @mapleFU @torwig

**Comment by git-hulk** [comment]
> I want to discuss how we can implement this feature. My idea is to count client memory within each worker (using `libevent`'s callback to count `evbuffer`) to avoid locking across workers. @AntiTopQuark @PragmaTwice @git-hulk @mapleFU @torwig

Yes, as mentioned in [previous comment](https://github.com/apache/kvrocks/pull/2498#issuecomment-2311996885). It would be better to limit the connection output first by checking the libevent buffer. And for the `maxmemory-clients`, I think we don't need to count every byte, just checking the libevent input and output buffer is good to me as well.

**Comment by AntiTopQuark** [comment]
> > I want to discuss how we can implement this feature. My idea is to count client memory within each worker (using `libevent`'s callback to count `evbuffer`) to avoid locking across workers. @AntiTopQuark @PragmaTwice @git-hulk @mapleFU @torwig
> 
> Yes, as mentioned in [previous comment](https://github.com/apache/kvrocks/pull/2498#issuecomment-2311996885). It would be better to limit the connection output first by checking the libevent buffer. And for the `maxmemory-clients`, I think we don't need to count every byte, just checking the libevent input and output buffer is good to me as well.

I have modified the code and moved the `evictionClients` function into `Worker`.
On another note, I believe that `maxmemory-clients` is used to limit the total size of connections. If you want to limit the size of the output buffer, you can use the client-output-buffer-limit configuration option. I will submit a PR for this in the future.

**Comment by mapleFU** [comment]
>  want to discuss how we can implement this feature. My idea is to count client memory within each worker (using libevent's callback to count evbuffer) to avoid locking across workers. 

Yeah I think holding a lock is harmful here, I can accept some untimely memory change, but global checking is harmful...

File: src/server/redis_connection.cc [hunk 2]
```diff
@@ -559,4 +562,37 @@ void Connection::ResetMultiExec() {
   DisableFlag(Connection::kMultiExec);
 }
 
+size_t Connection::GetConnectionMemoryUsed() const {
+  size_t total_memory = sizeof(*this);  // 包含所有成员变量的静态内存大小
```
Comment (PragmaTwice) [inline src/server/redis_connection.cc]: please dont include chinese comments.

File: src/server/redis_connection.cc [hunk 3]
```diff
@@ -559,4 +562,37 @@ void Connection::ResetMultiExec() {
...
+  for (const auto &channel : subscribe_shard_channels_) {
+    total_memory += channel.capacity();
+  }
+  for (const auto &cmd : multi_cmds_) {
+    total_memory += cmd.capacity();
+  }
+
+  if (saved_current_command_) {
+    total_memory += saved_current_command_->GetMemoryUsage();
+  }
```
Comment (PragmaTwice) [inline src/server/redis_connection.cc:593]: The code seems hard to maintain and also not like a precise estimation of memory usage.

I'm wondering if it's really necessary to limit the connection memory usage..

**Comment by PragmaTwice** [comment]
I believe implementing an elegant solution to track and restrict client connection memory usage is challenging and could reduce code maintainability. 

Instead, I suggest exploring alternative approaches, like limiting the number of clients.

File: src/server/redis_connection.cc (hunk 2 above)
Comment (AntiTopQuark) [reply src/server/redis_connection.cc]: done

**Comment by AntiTopQuark** [comment]
> I believe implementing an elegant solution to track and restrict client connection memory usage is challenging and could reduce code maintainability.
> 
> Instead, I suggest exploring alternative approaches, like limiting the number of clients.



> I believe implementing an elegant solution to track and restrict client connection memory usage is challenging and could reduce code maintainability.
> 
> Instead, I suggest exploring alternative approaches, like limiting the number of clients.

When I looked at Redis previously, it uses the maxclients configuration option to limit the number of clients, and maxmemory-clients to limit the total memory used by connections. If you only want to limit the output buffer, you can use client-output-buffer-limit to restrict the usage per connection.

**Comment by AntiTopQuark** [comment]
@PragmaTwice  @git-hulk  Hi, should we continue pushing this PR forward, or can it be closed?

**Pull request closed without merging**